pkg/storage/writer.go

```diff
@@ -2,11 +2,71 @@ package storage
 
 import (
 	"bytes"
+	"compress/gzip"
 	"context"
 )
 
+type interceptBuffer struct {
+	*bytes.Buffer
+	compressWriter *gzip.Writer
```

Comment:
make this an `interface { io.Writer; io.Closer; Flush() error }`, we may support other compression algorithms.